src/consensus/aft/raft.h

```diff
@@ -1756,43 +1771,61 @@ namespace aft
       const auto last_committable_idx = last_committable_index();
       const auto term_of_last_committable_idx =
         get_term_internal(last_committable_idx);
-
-      const auto answer =
+      const auto log_up_to_date =
         (r.term_of_last_committable_idx > term_of_last_committable_idx) ||
         ((r.term_of_last_committable_idx == term_of_last_committable_idx) &&
          (r.last_committable_idx >= last_committable_idx));
-
-      if (answer)
-      {
-        // If we grant our vote, we also acknowledge that an election is in
-        // progress.
-        restart_election_timeout();
-        leader_id.reset();
-        voted_for = from;
-      }
-      else
+      if (!log_up_to_date)
       {
-        RAFT_INFO_FMT(
-          "Voting against candidate at {}.{} because local state is at {}.{}",
+        RAFT_DEBUG_FMT(
+          "Request vote to {} from {}: candidate log {}.{} is not up-to-date "
+          "with ours {}.{}",
+          state->node_id,
+          from,
           r.term_of_last_committable_idx,
           r.last_committable_idx,
           term_of_last_committable_idx,
           last_committable_idx);
+        grant_vote = false;
+      }
+
+      if (grant_vote && r.election_type == ElectionType::RegularVote)
+      {
+        // If we grant our vote to a candidate, then an election is in progress
+        restart_election_timeout();
+        leader_id.reset();
+        voted_for = from;
       }
 
-      send_request_vote_response(from, answer);
+      RAFT_INFO_FMT(
+        "Request {}vote to {} from {}: {} vote to candidate at {}.{} with "
+        "local state at {}.{}",
+        r.election_type == ElectionType::PreVote ? "pre-" : "",
```

Comment:
It's quite easy to make enums JSON-printable, this is a little more code, but the log does look cleaner as a result.